### ISSUE TITLE
DLNA: The given key 'ObjectID' was not present in the dictionary

### ISSUE THREAD

**Issue by sam-cross** [jellyfin/jellyfin#4423]
**Describe the bug**
With DLNA set up, attempting to access the Jellyfin server through VLC UPnP or DLNA on a PlayStation 3 console results in an error, and the following is displayed in the log (with DLNA logging enabled):

`Emby.Dlna.ContentDirectory.ContentDirectoryService: Error processing control requestSystem.Collections.Generic.KeyNotFoundException: The given key 'ObjectID' was not present in the dictionary`

My media directory (`/media`) has `rwx` permissions for everyone. Jellyfin can read it fine and playback occurs in the browser and Android app, but DLNA does not work.

**System (please complete the following information):**
 - OS: Arch Linux x64 (in Proxmox x64)
 - Virtualization: Docker
 - Clients: VLC Media Player, PlayStation 3
 - Jellyfin Version: 10.7.0 (Docker image: jellyfin/jellyfin:2020-11-05.1)
 - Installed Plugins: None
 - Reverse Proxy: Nginx (*same issue occurs WITHOUT reverse proxy*)
 - Base URL: `/` or `/jellyfin/`
 - Networking: Docker Host
 - Storage: [e.g. local, NFS, cloud]

**To Reproduce**

1. Enable DLNA in Jellyfin with "Blast alive messages" enabled
2. Attempt to access Jellyfin DLNA server through VLC Media Player or PS3 media server browser
3. VLC infinitely tries to load directory, PS3 displays error code 7531
4. Below message appears (repeatedly) in log file

**Expected behavior**
Expected DLNA directory listing.

**Logs**
Below is the Jellyfin log content from scanning and attempting to browse the Jellyfin DLNA server:
```
[2020-11-06 12:29:17.572 +00:00] [ERR] [29] Emby.Dlna.ContentDirectory.ContentDirectoryService: Error processing control request
System.Collections.Generic.KeyNotFoundException: The given key 'ObjectID' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Emby.Dlna.ContentDirectory.ControlHandler.HandleBrowse(XmlWriter xmlWriter, IDictionary`2 sparams, String deviceId)
   at Emby.Dlna.ContentDirectory.ControlHandler.WriteResult(String methodName, IDictionary`2 methodParams, XmlWriter xmlWriter)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestInternalAsync(ControlRequest request)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestAsync(ControlRequest request)
[2020-11-06 12:29:17.739 +00:00] [ERR] [33] Emby.Dlna.ContentDirectory.ContentDirectoryService: Error processing control request
System.Collections.Generic.KeyNotFoundException: The given key 'ObjectID' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Emby.Dlna.ContentDirectory.ControlHandler.HandleBrowse(XmlWriter xmlWriter, IDictionary`2 sparams, String deviceId)
   at Emby.Dlna.ContentDirectory.ControlHandler.WriteResult(String methodName, IDictionary`2 methodParams, XmlWriter xmlWriter)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestInternalAsync(ControlRequest request)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestAsync(ControlRequest request)
[2020-11-06 12:29:17.977 +00:00] [ERR] [29] Emby.Dlna.ContentDirectory.ContentDirectoryService: Error processing control request
System.Collections.Generic.KeyNotFoundException: The given key 'ObjectID' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Emby.Dlna.ContentDirectory.ControlHandler.HandleBrowse(XmlWriter xmlWriter, IDictionary`2 sparams, String deviceId)
   at Emby.Dlna.ContentDirectory.ControlHandler.WriteResult(String methodName, IDictionary`2 methodParams, XmlWriter xmlWriter)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestInternalAsync(ControlRequest request)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestAsync(ControlRequest request)
[2020-11-06 12:29:18.406 +00:00] [ERR] [34] Emby.Dlna.ContentDirectory.ContentDirectoryService: Error processing control request
System.Collections.Generic.KeyNotFoundException: The given key 'ObjectID' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Emby.Dlna.ContentDirectory.ControlHandler.HandleBrowse(XmlWriter xmlWriter, IDictionary`2 sparams, String deviceId)
   at Emby.Dlna.ContentDirectory.ControlHandler.WriteResult(String methodName, IDictionary`2 methodParams, XmlWriter xmlWriter)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestInternalAsync(ControlRequest request)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestAsync(ControlRequest request)
[2020-11-06 12:29:19.235 +00:00] [ERR] [34] Emby.Dlna.ContentDirectory.ContentDirectoryService: Error processing control request
System.Collections.Generic.KeyNotFoundException: The given key 'ObjectID' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Emby.Dlna.ContentDirectory.ControlHandler.HandleBrowse(XmlWriter xmlWriter, IDictionary`2 sparams, String deviceId)
   at Emby.Dlna.ContentDirectory.ControlHandler.WriteResult(String methodName, IDictionary`2 methodParams, XmlWriter xmlWriter)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestInternalAsync(ControlRequest request)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestAsync(ControlRequest request)
[2020-11-06 12:29:20.866 +00:00] [ERR] [34] Emby.Dlna.ContentDirectory.ContentDirectoryService: Error processing control request
System.Collections.Generic.KeyNotFoundException: The given key 'ObjectID' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Emby.Dlna.ContentDirectory.ControlHandler.HandleBrowse(XmlWriter xmlWriter, IDictionary`2 sparams, String deviceId)
   at Emby.Dlna.ContentDirectory.ControlHandler.WriteResult(String methodName, IDictionary`2 methodParams, XmlWriter xmlWriter)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestInternalAsync(ControlRequest request)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestAsync(ControlRequest request)
[2020-11-06 12:29:24.095 +00:00] [ERR] [34] Emby.Dlna.ContentDirectory.ContentDirectoryService: Error processing control request
System.Collections.Generic.KeyNotFoundException: The given key 'ObjectID' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Emby.Dlna.ContentDirectory.ControlHandler.HandleBrowse(XmlWriter xmlWriter, IDictionary`2 sparams, String deviceId)
   at Emby.Dlna.ContentDirectory.ControlHandler.WriteResult(String methodName, IDictionary`2 methodParams, XmlWriter xmlWriter)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestInternalAsync(ControlRequest request)
   at Emby.Dlna.Service.BaseControlHandler.ProcessControlRequestAsync(ControlRequest request)
```

And here's a [Wireshark traffic capture](https://mega.nz/file/zAV1AKjS#AYvqo2DZXQdHNTHyWNAoaZ0vFOSR5dutJx-Sa5p5L6k) from my Windows 10 machine, running VLC and WMP, for good measure.

**Screenshots**
Jellyfin DLNA settings:
![image](https://user-images.githubusercontent.com/6699307/98366738-662ff180-202c-11eb-825c-5bc75350a6c9.png)

**Comment by sam-cross** [comment]
Interesting bit of info: I can initialise playback on Windows Media Player through the Jellyfin web interface, but it doesn't actually play any video. Windows Media Player does *not* list the Jellyfin server under its libraries, though.

VLC Media Player does not show up in the Cast menu on the Jellyfin web interface. The PS3 does *not* show up in the Cast menu either. PS3 can find the Jellyfin DLNA server, but cannot read from it - it sometimes displays error 2114 instead of the other code quoted above.

**Comment by PrplHaz4** [comment]
TLDR; For DLNA to work as of this post, you should be running 10.6.4 or unstable with #3994

It's important to note that you are on unstable - did this config work on 10.6.4 for you?

There are a lot of DLNA/network changes coming up - so if you're interested in running unstable and DLNA, you should be running a build with #3994 in it.

**Comment by sam-cross** [comment]
I'll try both 10.6.4 and unstable with that patch when I'm home and report
back!

DLNA wasn't working on 10.6.4 last I tried, however that's because I was
using a custom baseUrl, something which I later discovered broke DLNA
(#2377). I neglected to test without the baseUrl on that version.

**Comment by sam-cross** [comment]
Gone back to 10.6.4 now. Had a couple weird issues with the web interface (not loading, or sometimes just a black screen). I suspect a config issue but can't find anything in the Docker logs.

Anyway, ***it works***! So looks like the 20201105 unstable version does not support DLNA properly. I'd like to know what changed between the versions for it to break.

I'm going to try making a Docker image with #3994 in it, and I'll report back if that fixes (or breaks) anything.

**Comment by BaronGreenback** [comment]
Just checked the wireshark and ObjectID is definitely present.

```
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<s:Body>
  <u:Browse xmlns:u="urn:schemas-upnp-org:service:ContentDirectory:1">
    <ObjectID>0</ObjectID>
    <BrowseFlag>BrowseDirectChildren</BrowseFlag>
    <Filter>*</Filter>
    <StartingIndex>0</StartingIndex>
    <RequestedCount>5000</RequestedCount>
    <SortCriteria></SortCriteria>
  </u:Browse>
</s:Body>
</s:Envelope>
```

**Comment by sam-cross** [comment]
@BaronGreenback  I haven't done a Wireshark capture of the working DLNA server yet, but I'll get that done when I can to compare.

If ObjectID is present, I'm not sure what the error is. Have you got any ideas for troubleshooting?

I'm still trying to get your networking2 patch built so I can try that, but I'm getting a Mono build error (something about needing to specify a project file?) so that's not happening right now.

**Comment by BaronGreenback** [comment]
I've located where it's crashing - and i know why - because it didn't parse the soap request correctly for some reason.

Just checking the dlna server against UPnP Developer tools. It has a nice program that checks the validity of device responses.

**Comment by crobibero** [comment]
This PR introduced a bug where the headers aren’t set on the response object
https://github.com/jellyfin/jellyfin/pull/4212

**Comment by BaronGreenback** [comment]
Have reproduced here - will submit a PR

**Comment by BaronGreenback** [comment]
return result; is missing from line 172.

**Comment by BaronGreenback** [comment]
@sam-cross Thanks for your help.
